mlflow/transformers.py

```diff
@@ -0,0 +1,789 @@
+    card_data = model_card if model_card is not None else _fetch_model_card(transformers_model)
+
+    # If the card data can be acquired, save the text and the data separately
+    if card_data:
+        write_to(str(path.joinpath(_CARD_TEXT_FILE_NAME)), card_data.text)
+        with path.joinpath(_CARD_DATA_FILE_NAME).open("w") as file:
+            yaml.safe_dump(card_data.data.to_dict(), stream=file, default_flow_style=False)
+
+    model_bin_kwargs = {_PIPELINE_BINARY_KEY: _PIPELINE_BINARY_FILE_NAME}
+
+    # Only allow a subset of task types to have a pyfunc definition.
+    # Currently supported types are NLP-based language tasks which have a pipeline definition
+    # consisting exclusively of a Model and a Tokenizer.
+    if _should_add_pyfunc_to_model(built_pipeline):
+        pyfunc.add_to_model(
+            mlflow_model,
+            loader_module="mlflow.transformers",
+            conda_env=_CONDA_ENV_FILE_NAME,
+            python_env=_PYTHON_ENV_FILE_NAME,
+            code=code_dir_subpath,
+            **model_bin_kwargs,
+        )
+    flavor_conf.update(**model_bin_kwargs)
+    mlflow_model.add_flavor(
+        FLAVOR_NAME,
+        transformers_version=transformers.__version__,
+        code=code_dir_subpath,
+        **flavor_conf,
+    )
+    mlflow_model.save(str(path.joinpath(MLMODEL_FILE_NAME)))
+
+    if conda_env is None:
+        if pip_requirements is None:
+            if isinstance(transformers_model, transformers.Pipeline):
+                default_reqs = get_default_pip_requirements(transformers_model.model)
+            else:
+                default_reqs = get_default_pip_requirements(transformers_model[_MODEL_KEY])
+            inferred_reqs = infer_pip_requirements(str(path), FLAVOR_NAME, fallback=default_reqs)
+            default_reqs = sorted(set(inferred_reqs).union(default_reqs))
+        else:
+            default_reqs = None
+        conda_env, pip_requirements, pip_constraints = _process_pip_requirements(
+            default_reqs, pip_requirements, extra_pip_requirements
+        )
+    else:
+        conda_env, pip_requirements, pip_constraints = _process_conda_env(conda_env)
+
+    with path.joinpath(_CONDA_ENV_FILE_NAME).open("w") as f:
+        yaml.safe_dump(conda_env, stream=f, default_flow_style=False)
+
+    if pip_constraints:
+        write_to(str(path.joinpath(_CONSTRAINTS_FILE_NAME)), "\n".join(pip_constraints))
+
+    write_to(str(path.joinpath(_REQUIREMENTS_FILE_NAME)), "\n".join(pip_requirements))
+
+    _PythonEnv.current().to_yaml(str(path.joinpath(_PYTHON_ENV_FILE_NAME)))
+
+
+@experimental
+@format_docstring(LOG_MODEL_PARAM_DOCS.format(package_name=FLAVOR_NAME))
+def log_model(
+    transformers_model,
+    artifact_path: str,
+    processor=None,
+    task: Optional[str] = None,
+    model_card=None,
+    conda_env: Optional[
+        Union[str, Dict[str, Union[str, List[Union[str, Dict[str, List[str]]]]]]]
```

Comment:
This type hint is still pretty intense. Can we remove it?

Reply:
sorry. missed this one. Removed!